### DISCUSSION TITLE
Include Contacts next to their phone number

### DISCUSSION THREAD

**Discussion by oshionniemela** [ReagentX/imessage-exporter#289] [Q&A] [answered]
---

### iMessage Exporter Enhancement: Contact Name Integration

iMessage Exporter is incredibly useful, but I found it lacking one feature: listing the contact's name next to their number (both in the file name and within the HTML messages). To address this, I wrote a script that accomplishes this task, and I believe it could be beneficial for others. Perhaps it could even be integrated into iMessage Exporter itself. I suggest adding a `--contacts {contact-vcf-location}` option for this purpose. For example:

```bash
$ imessage-exporter -f html -c compatible --contacts {contact-vcf-location} --country-code US
```

For now, to use this script separately, follow these steps:

#### Steps to Use the Contact Parser Script

1. **Export Your Contacts to a VCF File**
   - On your Mac, open the Contacts app.
   - Select all contacts.
   - Navigate to `File` -> `Export` -> `Export vCard`.

2. **Prepare the Contact Parser Script**
   - Open the terminal and navigate to the `Contact Parser` directory.
   - Make the script executable with:
     ```bash
     chmod +x run_contact_parser.sh
     ```
   - Run the script with:
     ```bash
     ./run_contact_parser.sh {contacts-vcf-location} {imessages-location}
     ```
   - This will update every HTML message file with the contact's name (if found in the VCF file).

3. **Handling International Numbers**
   - By default, the script assumes US numbers. If you are in another country, you can pass a third parameter with your 2-letter country code:
     ```bash
     ./run_contact_parser.sh {contacts-vcf-location} {imessages-location} {country-code}
     ```
   - For example:
     ```bash
     ./run_contact_parser.sh contacts.vcf imessages-directory UK
     ```

#### Important Notes

- **Backup Your Data**: Before running the script, please ensure you have backed up your exported HTML messages.
- **Tested on Mac**: This script has only been tested on a Mac environment.

Here is the repository link: https://github.com/oshionniemela/imessage-exporter-name-enhancer

I hope it proves helpful! Let me know if you have any questions or feedback.

---

**Comment by ReagentX** [comment]
Please don't post a zip for people to download; we don't know what's inside. If you want to share the code you should publish a repository or a [gist](https://gist.github.com/).

**Reply by oshionniemela** [reply]
Oh sorry about that I'll add it as a repository then and then share the link.

**Reply by ReagentX** [reply]
Cheers, this is cool work.

**Comment by oshionniemela** [comment]
Here is the repository link: https://github.com/oshionniemela/imessage-exporter-name-enhancer

**Comment by MMachado05** [comment]
This seems pretty cool, but I definitely think it's limited by the fact that you need access to a mac in order for it to run. A big sell of iMessage Exporter's is that it's platform agnostic; so long as you have either an iMessage database or an unencrypted iOS backup available, you can run the export anywhere. For example, I use Linux and was able to use this project super easily.

Unless this feature is integrated with iMessage Exporter in such a way that it can use the unencrypted backup to find contact information (not sure how that works), I don't think it would be appropriate to include this script in its current state in the project seeing as it goes against the platform agnosticism the project currently provides.

**Reply by oshionniemela** [reply]
Hi @MMachado05, I'm sorry for the delay in responding, I'm just seeing these comments. 

I’m on a Mac and wasn’t aware that iMessage Exporter works across platforms, which is nice. I’m curious how one can get their iMessage database without using a Mac. If you can get the database without a Mac, you should also be able to export your contacts to a VCF file, as VCF files aren’t Mac-specific. I haven’t tested this script on Windows or Linux, so there might be minor syntax issues (e.g., path separators), but I believe Python scripts should run on Windows from their Terminal/PowerShell, so it should still work.

The script simply goes through each contact in a VCF file and checks if each contact's phone number appears in any HTML file generated by iMessage Exporter. If found, it prepends the contact’s name to the number. This process isn’t platform-specific—once you have your contacts exported as a VCF file, it should work.

So just like you provide your iMessage database, if you provide your contacts as a VCF file I think it should work.

**Reply by MMachado05** [reply]
> I'm curious how one can get their iMessage database without using a Mac.

In my case, I actually used an unencrypted full iPhone backup to get my messages exported, as iMessage-Exported supports this source. Doing that is as simple as using something like [libimobiledevice](https://github.com/libimobiledevice/libimobiledevice) or [pymobiledevice3](https://github.com/doronz88/pymobiledevice3) and running the associated command with an iPhone plugged into your Linux system (though the latter is also platform-agnostic, which is nice). 

I'm not sure, personally, how to get the VCF file for your contacts from an unencrypted iPhone backup, but as you say, it might be a relatively straightforward process! I'm not one to say though.

**Reply by ReagentX** [reply]
In addition to those projects, it is also worth noting that you can backup an iPhone using Finder or the new [Apple Devices](https://support.apple.com/guide/devices-windows/install-the-apple-devices-app-mchl5ded2763/windows) app on Windows.

**Reply by oshionniemela** [reply]
@MMachado05 @ReagentX Good to know! Also, I found some other ways to export contacts as a vCard ('.vcf' file) without a Mac:

- Using iCloud.com:
  - Sync your iPhone contacts to iCloud.
  - Log in to [iCloud Contacts](https://www.icloud.com/contacts).
  - Select all contacts by pressing Cmd+A (Mac) or Ctrl+A (Windows).
  - Click the share icon (top-right corner).
  - Choose "Export vCard" to download the file to your computer.

- From the iOS Contacts app:
  - Open the Contacts app.
  - Navigate to Lists (top-left) and tap and hold a specific list.
  - Choose Export from the options.
  - Confirm which details you want to include and tap Done.
  - Select your preferred sharing method such as emailing yourself the vcf file.
  - Open that email from your computer and you can then download it.

**Reply by oshionniemela** [reply]
I also wanted to mention that my contact_parser.py script is cross-platform and should run on any machine, as it's written in Python. However, it does require two third-party packages, vobject and phonenumbers, to function correctly. To simplify the setup, I created a run_contact_parser.sh script for macOS/Linux that would automatically install these dependencies in a temporary virtual environment, run the script, and clean up afterward. However, I just had ChatGPT help me create a Windows-compatible .bat script to achieve the same functionality. So this should now work on any operating system.

**Comment by e40** [comment]
I find this incredibly useful and I am very appreciative of @oshionniemela's work.  Thank you.

I hope this functionality is included in this repo, even if it requires macOS.  A very significant percentage of iOS users also are macOS users.

**Reply by oshionniemela** [reply]
Thanks @e40 I appreciate it!

**Comment by loushin47** [comment]
Thank you @oshionniemela for your contribution to this app.  I got it working on a Windows PC.  Thank you!

**Reply by oshionniemela** [reply]
Thanks @loushin47, glad it worked for you!

**Comment by tdoyle3** [comment]
Thank you @oshionniemela - this was exactly what I was looking for! Have you thought about adding a routine to rename multiple contact files with each of the names?

**Reply by oshionniemela** [reply]
Thanks @tdoyle3! I decided not to include everyone's names in a group chat in the file name because some group chats can have many people which would make the file names quite long, so I just decided to only include the phone numbers.

**Reply by memphis-10** [reply]
HI! Would you consider, for group chat file names, listing each contact name - but limiting the file name to 50 characters (so it just stops after 50, then adds the extension)?  Or give that option?  This is how a paid exporter I work with does it.

**Comment by memphis-10** [comment]
Hello! This is a brilliant addition- thank you for sharing! Maybe they can implement this as an option in the overall program - that would be great!

In the meantime, I have a couple questions:
1) Will this change all of these contact number/name instances: html file, txt file, AND actual file name?
2) Is it possible to *replace* the phone number with the contact name? Or give us the option?

These would be really huge, organizationally.

Thanks again! Excited to try this out!

**Reply by memphis-10** [reply]
Also, if I am US-based, but some of my iMessages are from other countries, how is that handled?  Thanks!

**Reply by ReagentX** [reply]
To provide feedback on the other repository, open a discussion or issue there; not in this thread.

**Reply by memphis-10** [reply]
OK

**Comment by ReagentX** [comment, marked as answer]
Native support for contact detail resolution was merged via #631 and was released in [Desert Globemallow](https://github.com/ReagentX/imessage-exporter/releases/tag/3.2.0).

I am going to close this discussion as it is no longer required.
